### PR TITLE
when nimvm stmt

### PR DESCRIPTION
**Pull request by yglukhov**

### CODE REVIEW

**Comment by Varriount** [comment]
When would this be used?

**Comment by yglukhov** [comment]
To differentiate execution paths between nim vm and runtime. E.g.
```nim
proc someProcThatMayRunInCompileTime(): bool =
  when nimvm:
    result = true
  else:
    when defined(js):
      {.emit: "`result` = false;".}
    else
      result = false

const ctVal = someProcThatMayRunInCompileTime()
assert(ctVal == true)

let rtVal = someProcThatMayRunInCompileTime()
assert(rtVal == false)
```

**Comment by nanoant** [comment]
@yglukhov Makes sense, but can't you use `define(vm)` instead introducing new keyword? It will make whole `when ...` more consistent

**Comment by yglukhov** [comment]
@nanoant, that's a matter of taste, imho. Anyways this nimvm stuff makes all the when node behave differently. So to me, nimvm seems more unique than defined(vm), and it's good because it's kinda hinting that some magic is happening here. Besides, nimvm is not a keyword in it's common sense. E.g. you may shadow it.

File: compiler/ast.nim [hunk 1]
```diff
@@ -647,7 +648,7 @@ const
   # thus cannot be overloaded (also documented in the spec!):
   SpecialSemMagics* = {
     mDefined, mDefinedInScope, mCompiles, mLow, mHigh, mSizeOf, mIs, mOf,
-    mShallowCopy, mExpandToAst, mParallel, mSpawn, mAstToStr}
+    mShallowCopy, mExpandToAst, mParallel, mSpawn, mAstToStr, mNimvm }
```
Comment (Araq) [inline compiler/ast.nim]: Why? It can be overloaded just fine except it's a let variable.

File: compiler/semparallel.nim [hunk 2]
```diff
@@ -364,6 +364,7 @@ proc analyse(c: var AnalysisCtx; n: PNode) =
         if not isSpawned: analyse(c, value)
   of nkCaseStmt: analyseCase(c, n)
   of nkIfStmt, nkIfExpr: analyseIf(c, n)
+  of nkWhen: echo "SEMPARALLEL"; analyseIf(c, n)
```
Comment (Araq) [inline compiler/semparallel.nim]: debug output remaining.

File: compiler/ast.nim [hunk 3]
```diff
@@ -1569,6 +1574,11 @@ proc isEmptyType*(t: PType): bool {.inline.} =
   ## 'void' and 'stmt' types are often equivalent to 'nil' these days:
   result = t == nil or t.kind in {tyEmpty, tyStmt}
 
+proc isWhenNimvmStmt*(n: PNode): bool =
+  n.kind == nkWhen and n.sons.len == 2 and n.sons[0].kind == nkElifBranch and
+      n.sons[1].kind == nkElse and n.sons[0].sons[0].kind == nkIdent and
+      n.sons[0].sons[0].ident.s == "nimvm"
```
Comment (Araq) [inline compiler/ast.nim]: Still wrong, you need to check for the symbol.magic == mNimvm

File: compiler/ast.nim [hunk 4]
```diff
@@ -1472,22 +1473,26 @@ proc hasNilSon*(n: PNode): bool =
...
-  of nkEmpty..nkNilLit: result = n.kind == kind
-  else:
-    for i in countup(0, sonsLen(n) - 1):
-      if (n.sons[i].kind == kind) or hasSubnodeWith(n.sons[i], kind):
-        return true
-    result = false
+  result = n.containsNode(proc(cn: PNode): bool =
+    cn.kind == kind)
+
+proc hasSubnodeWith*(n: PNode, magic: TMagic): bool =
```
Comment (Araq) [inline compiler/ast.nim]: You never use this and hasSubnodeWith shouldn't call a callback for every node, it's supposed to be very fast.

**Comment by yglukhov** [comment]
All fixed.

File: compiler/semexprs.nim [hunk 5]
```diff
@@ -1764,22 +1764,39 @@ proc semWhen(c: PContext, n: PNode, semCheck = true): PNode =
...
     else: result = e
 
+  # Check if the node is "when nimvm"
+  # when nimvm:
+  #   ...
+  # else:
+  #   ...
+  let whenNimvm = n.sons.len == 2 and n.sons[0].kind == nkElifBranch and
+      n.sons[1].kind == nkElse and n.sons[0].sons[0].kind == nkIdent and
+      lookUp(c, n.sons[0].sons[0]).magic == mNimvm
```
Comment (Araq) [inline compiler/semexprs.nim]: Er, so 'when nimvm' **requires** an else? Hrm, makes sense I guess. But 'nimvm' needs to be documented in the manual's when section too.

**Comment by yglukhov** [comment]
Docs updated.

**Pull request merged**